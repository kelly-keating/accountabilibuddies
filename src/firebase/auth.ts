import { 
  getAuth, 
  createUserWithEmailAndPassword, 
  signInWithEmailAndPassword,
  signOut
} from "firebase/auth"

const auth = getAuth()
export default auth

export function getUserId (): string | null {
  return auth.currentUser?.uid || null
}

export function register (email: string, password: string, errFn: (code: string) => void) {
  return createUserWithEmailAndPassword(auth, email, password)
    .then((userCredential) => userCredential.user)
    .catch(({ code }) => errFn(readError(code)))
}

export function login (email: string, password: string, errFn: (code: string) => void) {
  return signInWithEmailAndPassword(auth, email, password)
    .then((userCredential) => userCredential.user)
    .catch(({ code }) => errFn(readError(code)))
}

export function logout () {
  return signOut(auth)
    .then(() => {
      console.log('LOGGED OUT')
    }).catch((error) => {
      console.log('LOG OUT ERROR:', error.code, error.message)
    })
}

function readError (code: string) {
  console.log(code)
  switch (code) {
    case 'auth/weak-password':
      return 'Password too weak'
    case 'auth/email-already-in-use':
      return 'Email already registered'
    case 'auth/wrong-password':
      return 'Password incorrect'
    case 'auth/user-not-found':
      return 'Email incorrect'
    case 'auth/too-many-requests':
      return 'Too many requests - please try again later'
    default:
      return code
  }
}
