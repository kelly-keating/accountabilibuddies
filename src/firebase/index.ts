import { initializeApp } from "firebase/app"

const firebaseConfig = {
  apiKey: "AIzaSyAmQ31pldBDEXR6Hcbr4b5BJqeZEYelKMo",
  authDomain: "accountabilibuddiez.firebaseapp.com",
  projectId: "accountabilibuddiez",
  storageBucket: "accountabilibuddiez.appspot.com",
  messagingSenderId: "836544601261",
  appId: "1:836544601261:web:ca1b8860a3f48a44ffba24",
}

export const firebaseApp = initializeApp(firebaseConfig)
