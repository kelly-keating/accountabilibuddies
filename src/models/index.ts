import * as FirebaseAuth from 'firebase/auth'

export type FirebaseUser = FirebaseAuth.User
export interface User {
  displayName: string
  photoUrl: string
  lastEntry: string | null
  ratings: Record<string, RatingKeys>
}

export type AllUsers = Record<string, User>
export type ActiveUsers = Record<string, boolean>

export interface RatingKeys {
  id: string
  text: string
  current: boolean
  col?: string
}
export type Ratings = Record<string, Record<string, Record<string, number>>>

export type UpdateMode = "displayOnly" | "update" | "delete"
