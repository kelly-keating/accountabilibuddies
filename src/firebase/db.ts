import { FirebaseUser } from '../models'
import { getDatabase, ref, set, get, remove } from 'firebase/database'

import { getUserId } from './auth'
import { getNextWednesday, getThisWednesday } from '../dateUtils'

const db = getDatabase()
export default db

// USER

export async function checkUserExists(uid: string) {
  const userInfoRef = ref(db, `users/` + uid)
  try {
    const snapshot = await get(userInfoRef)
    return snapshot.exists()
  } catch (error) {
    console.error(error)
  }
}

export function addNewUser(userDetails: FirebaseUser) {
  const uid = getUserId() || userDetails.uid
  const newUserData = {
    displayName: userDetails.displayName,
    photoUrl: userDetails.photoURL,
  }

  const userRef = ref(db, `users/` + uid)
  set(userRef, newUserData)

  setUserActive(uid)
}

// RATINGS

export function addNewUserRating(text: string) {
  const uid = getUserId()
  const uniqueId = `${Date.now()}-${Math.random().toString(36).substring(2, 9)}`

  const ratingRef = ref(db, `users/${uid}/ratings/` + uniqueId)
  const ratingObj = {
    id: uniqueId,
    current: true,
    text,
  }

  set(ratingRef, ratingObj)
}

export function updateRatingText(id: string, newText: string) {
  const uid = getUserId()
  const ratingLabelRef = ref(db, `users/${uid}/ratings/${id}/text`)

  set(ratingLabelRef, newText)
}

export function updateRatingActive(id: string, newState: boolean) {
  const uid = getUserId()
  const ratingLabelRef = ref(db, `users/${uid}/ratings/${id}/current`)

  set(ratingLabelRef, newState)
}

export function deleteRatingForever(id: string) {
  const uid = getUserId()
  const ratingLabelRef = ref(db, `users/${uid}/ratings/` + id)

  remove(ratingLabelRef)
}

export function updateRatingThisWeek(ratingId: string, val: number) {
  const uid = getUserId()
  const date = getThisWednesday()
  const ratingRef = ref(db, `ratings/${uid}/${date}/${ratingId}`)

  set(ratingRef, val)
}

// GOALS

export function addNewGoal(text: string) {
  const uid = getUserId()
  const date = getNextWednesday()
  const goalRef = ref(db, `goals/${uid}/${date}`)

  const newGoal = {
    text,
    date,
  }

  set(goalRef, newGoal)
}

export function updateGoalText(text: string) {
  const uid = getUserId()
  const date = getNextWednesday()
  const goalRef = ref(db, `goals/${uid}/${date}/text`)

  set(goalRef, text)
}

export function setGoalComplete(date: string, status: boolean) {
  const uid = getUserId()
  const goalRef = ref(db, `goals/${uid}/${date}/completed`)

  set(goalRef, status)
}

// ACTIVE USERS

export function setUserActive(uid: string) {
  const activeRef = ref(db, `activeUsers/` + uid)
  set(activeRef, true)
}
