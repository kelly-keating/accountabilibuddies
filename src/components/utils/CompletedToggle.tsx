import { ButtonGroup, IconButton } from '@chakra-ui/react'
import { CheckIcon, CloseIcon } from '@chakra-ui/icons'
import { Goal } from '../../models'

import { setGoalComplete } from '../../firebase/db'
import { showAsIncomplete } from '../../colourUtils'

interface Props {
  goal: Goal
}

// TODO: don't put this in a Text tag
function CompletedToggle({ goal }: Props) {
  return (
    <ButtonGroup size="sm" isAttached variant="solid" mr="5px">
      <IconButton
        aria-label="mark completed"
        icon={<CheckIcon />}
        onClick={() => setGoalComplete(goal.date, true)}
        colorScheme={goal.completed ? 'green' : 'gray'}
        isDisabled={goal.completed}
      />
      <IconButton
        aria-label="mark incomplete"
        icon={<CloseIcon />}
        onClick={() => setGoalComplete(goal.date, false)}
        colorScheme={showAsIncomplete(goal) ? 'red' : 'gray'}
        isDisabled={!goal.completed}
      />
    </ButtonGroup>
  )
}

export default CompletedToggle
