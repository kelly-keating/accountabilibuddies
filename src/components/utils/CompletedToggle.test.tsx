import userEvent from '@testing-library/user-event'

import { render, screen } from '@testing-library/react'
import { describe, it, expect, vi } from 'vitest'
import CompletedToggle from './CompletedToggle'
import { setGoalComplete } from '../../firebase/db'
import { Goal } from '../../models'

vi.mock('../../firebase/db', () => ({
  setGoalComplete: vi.fn(),
}))

vi.mock('../../colourUtils', () => ({
  showAsIncomplete: vi.fn((g: Goal) => !g.completed),
}))

describe('CompletedToggle Component', () => {
  describe('renders buttons with correct state', () => {
    it('when goal is incomplete', () => {
      const mockGoal: Goal = {
        date: '2024-01-01',
        completed: false,
        text: 'Test goal',
      }
      const { container } = render(<CompletedToggle goal={mockGoal} />)
      expect(container).toMatchSnapshot()

      const markCompleteButton = screen.getByLabelText('mark completed')
      expect(markCompleteButton).toBeInTheDocument()
      expect(markCompleteButton).toBeEnabled()

      const markIncompleteButton = screen.getByLabelText('mark incomplete')
      expect(markIncompleteButton).toBeInTheDocument()
      expect(markIncompleteButton).toBeDisabled()
    })

    it('when goal is complete', () => {
      const mockGoal: Goal = {
        date: '2024-01-01',
        completed: true,
        text: 'Test goal',
      }

      const { container } = render(<CompletedToggle goal={mockGoal} />)
      expect(container).toMatchSnapshot()

      const markCompleteButton = screen.getByLabelText('mark completed')
      expect(markCompleteButton).toBeInTheDocument()
      expect(markCompleteButton).toBeDisabled()

      const markIncompleteButton = screen.getByLabelText('mark incomplete')
      expect(markIncompleteButton).toBeInTheDocument()
      expect(markIncompleteButton).toBeEnabled()
    })

    it('when completion is unknown', () => {
      const mockGoal: Goal = {
        date: '2024-01-01',
        text: 'Test goal',
      }

      const { container } = render(<CompletedToggle goal={mockGoal} />)
      expect(container).toMatchSnapshot()

      const markCompleteButton = screen.getByLabelText('mark completed')
      expect(markCompleteButton).toBeInTheDocument()
      expect(markCompleteButton).toBeEnabled()

      const markIncompleteButton = screen.getByLabelText('mark incomplete')
      expect(markIncompleteButton).toBeInTheDocument()
      expect(markIncompleteButton).toBeDisabled()
    })

    it('when goal data changes', () => {
      const mockGoal: Goal = {
        date: '2024-01-01',
        text: 'Test goal',
        completed: false,
      }

      const { container, rerender } = render(
        <CompletedToggle goal={mockGoal} />,
      )
      expect(container).toMatchSnapshot()
      const markCompleteButton = screen.getByLabelText('mark completed')
      const markIncompleteButton = screen.getByLabelText('mark incomplete')

      expect(markCompleteButton).toBeEnabled()
      expect(markIncompleteButton).toBeDisabled()

      mockGoal.completed = true
      rerender(<CompletedToggle goal={mockGoal} />)

      expect(markCompleteButton).not.toBeEnabled()
      expect(markIncompleteButton).not.toBeDisabled()

      mockGoal.completed = false
      rerender(<CompletedToggle goal={mockGoal} />)

      expect(markCompleteButton).toBeEnabled()
      expect(markIncompleteButton).toBeDisabled()
    })
  })

  describe('calls setGoalComplete', () => {
    it('when mark complete is clicked', async () => {
      const user = userEvent.setup()
      const mockGoal: Goal = {
        date: '2024-01-01',
        completed: false,
        text: 'Test goal',
      }

      render(<CompletedToggle goal={mockGoal} />)

      const markCompleteButton = screen.getByLabelText('mark completed')

      await user.click(markCompleteButton)
      expect(setGoalComplete).toHaveBeenCalledWith(mockGoal.date, true)
      expect(setGoalComplete).toHaveBeenCalledTimes(1)

      await user.click(markCompleteButton)
      expect(setGoalComplete).toHaveBeenCalledTimes(2)
    })

    it('when mark incomplete is clicked', async () => {
      const user = userEvent.setup()
      const mockGoal: Goal = {
        date: '2024-01-01',
        completed: true,
        text: 'Test goal',
      }

      const { container } = render(<CompletedToggle goal={mockGoal} />)
      expect(container).toMatchSnapshot()

      const markIncompleteButton = screen.getByLabelText('mark incomplete')

      await user.click(markIncompleteButton)
      await user.click(markIncompleteButton)
      await user.click(markIncompleteButton)

      expect(setGoalComplete).toHaveBeenCalledWith(mockGoal.date, false)
      expect(setGoalComplete).toHaveBeenCalledTimes(3)
    })

    it('only when enabled buttons are clicked', async () => {
      const user = userEvent.setup()
      const mockGoal: Goal = {
        date: '2024-01-01',
        completed: false,
        text: 'Test goal',
      }

      const { container } = render(<CompletedToggle goal={mockGoal} />)
      expect(container).toMatchSnapshot()

      const markCompleteButton = screen.getByLabelText('mark completed')
      const markIncompleteButton = screen.getByLabelText('mark incomplete')

      await user.click(markIncompleteButton)
      await user.click(markCompleteButton)
      await user.click(markIncompleteButton)
      await user.click(markIncompleteButton)

      expect(setGoalComplete).toHaveBeenCalledWith(mockGoal.date, true)
      expect(setGoalComplete).toHaveBeenCalledTimes(1)
    })
  })
})
