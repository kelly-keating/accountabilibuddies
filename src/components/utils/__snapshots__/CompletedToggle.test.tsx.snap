// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CompletedToggle Component > calls setGoalComplete > only when enabled buttons are clicked 1`] = `
<div>
  <div
    class="chakra-button__group css-1a99rsg"
    data-attached=""
    data-orientation="horizontal"
    role="group"
  >
    <button
      aria-label="mark completed"
      class="chakra-button css-24qtlu"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="chakra-icon css-onkibi"
        focusable="false"
        viewBox="0 0 14 14"
      >
        <g
          fill="currentColor"
        >
          <polygon
            points="5.5 11.9993304 14 3.49933039 12.5 2 5.5 8.99933039 1.5 4.9968652 0 6.49933039"
          />
        </g>
      </svg>
    </button>
    <button
      aria-label="mark incomplete"
      class="chakra-button css-24qtlu"
      disabled=""
      type="button"
    >
      <svg
        aria-hidden="true"
        class="chakra-icon css-onkibi"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`CompletedToggle Component > calls setGoalComplete > when mark incomplete is clicked 1`] = `
<div>
  <div
    class="chakra-button__group css-1a99rsg"
    data-attached=""
    data-orientation="horizontal"
    role="group"
  >
    <button
      aria-label="mark completed"
      class="chakra-button css-24qtlu"
      disabled=""
      type="button"
    >
      <svg
        aria-hidden="true"
        class="chakra-icon css-onkibi"
        focusable="false"
        viewBox="0 0 14 14"
      >
        <g
          fill="currentColor"
        >
          <polygon
            points="5.5 11.9993304 14 3.49933039 12.5 2 5.5 8.99933039 1.5 4.9968652 0 6.49933039"
          />
        </g>
      </svg>
    </button>
    <button
      aria-label="mark incomplete"
      class="chakra-button css-24qtlu"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="chakra-icon css-onkibi"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`CompletedToggle Component > renders buttons with correct state > when completion is unknown 1`] = `
<div>
  <div
    class="chakra-button__group css-1a99rsg"
    data-attached=""
    data-orientation="horizontal"
    role="group"
  >
    <button
      aria-label="mark completed"
      class="chakra-button css-24qtlu"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="chakra-icon css-onkibi"
        focusable="false"
        viewBox="0 0 14 14"
      >
        <g
          fill="currentColor"
        >
          <polygon
            points="5.5 11.9993304 14 3.49933039 12.5 2 5.5 8.99933039 1.5 4.9968652 0 6.49933039"
          />
        </g>
      </svg>
    </button>
    <button
      aria-label="mark incomplete"
      class="chakra-button css-24qtlu"
      disabled=""
      type="button"
    >
      <svg
        aria-hidden="true"
        class="chakra-icon css-onkibi"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`CompletedToggle Component > renders buttons with correct state > when goal data changes 1`] = `
<div>
  <div
    class="chakra-button__group css-1a99rsg"
    data-attached=""
    data-orientation="horizontal"
    role="group"
  >
    <button
      aria-label="mark completed"
      class="chakra-button css-24qtlu"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="chakra-icon css-onkibi"
        focusable="false"
        viewBox="0 0 14 14"
      >
        <g
          fill="currentColor"
        >
          <polygon
            points="5.5 11.9993304 14 3.49933039 12.5 2 5.5 8.99933039 1.5 4.9968652 0 6.49933039"
          />
        </g>
      </svg>
    </button>
    <button
      aria-label="mark incomplete"
      class="chakra-button css-24qtlu"
      disabled=""
      type="button"
    >
      <svg
        aria-hidden="true"
        class="chakra-icon css-onkibi"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`CompletedToggle Component > renders buttons with correct state > when goal is complete 1`] = `
<div>
  <div
    class="chakra-button__group css-1a99rsg"
    data-attached=""
    data-orientation="horizontal"
    role="group"
  >
    <button
      aria-label="mark completed"
      class="chakra-button css-24qtlu"
      disabled=""
      type="button"
    >
      <svg
        aria-hidden="true"
        class="chakra-icon css-onkibi"
        focusable="false"
        viewBox="0 0 14 14"
      >
        <g
          fill="currentColor"
        >
          <polygon
            points="5.5 11.9993304 14 3.49933039 12.5 2 5.5 8.99933039 1.5 4.9968652 0 6.49933039"
          />
        </g>
      </svg>
    </button>
    <button
      aria-label="mark incomplete"
      class="chakra-button css-24qtlu"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="chakra-icon css-onkibi"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`CompletedToggle Component > renders buttons with correct state > when goal is incomplete 1`] = `
<div>
  <div
    class="chakra-button__group css-1a99rsg"
    data-attached=""
    data-orientation="horizontal"
    role="group"
  >
    <button
      aria-label="mark completed"
      class="chakra-button css-24qtlu"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="chakra-icon css-onkibi"
        focusable="false"
        viewBox="0 0 14 14"
      >
        <g
          fill="currentColor"
        >
          <polygon
            points="5.5 11.9993304 14 3.49933039 12.5 2 5.5 8.99933039 1.5 4.9968652 0 6.49933039"
          />
        </g>
      </svg>
    </button>
    <button
      aria-label="mark incomplete"
      class="chakra-button css-24qtlu"
      disabled=""
      type="button"
    >
      <svg
        aria-hidden="true"
        class="chakra-icon css-onkibi"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</div>
`;
